pkg/api/handlers/oauth/token.go

```diff
@@ -0,0 +1,216 @@
+package oauth
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/obot-platform/obot/apiclient/types"
+	"github.com/obot-platform/obot/logger"
+	"github.com/obot-platform/obot/pkg/api"
+	v1 "github.com/obot-platform/obot/pkg/storage/apis/obot.obot.ai/v1"
+	"github.com/obot-platform/obot/pkg/storage/selectors"
+	"golang.org/x/crypto/bcrypt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var log = logger.Package()
+
+func (h *handler) token(req api.Context) error {
+	if err := req.ParseForm(); err != nil {
+		return err
+	}
+
+	clientName := req.FormValue("client_id")
+	if clientName == "" {
+		return types.NewErrBadRequest("%v", Error{
+			Code:        ErrInvalidRequest,
+			Description: "client_id is required",
+		})
+	}
+
+	clientNamespace, clientName, ok := strings.Cut(clientName, ":")
+	if !ok {
+		return types.NewErrBadRequest("%v", Error{
+			Code:        ErrInvalidRequest,
+			Description: "client_id is invalid",
+		})
+	}
+
+	var client v1.OAuthClient
+	if err := req.Storage.Get(req.Context(), kclient.ObjectKey{Namespace: clientNamespace, Name: clientName}, &client); err != nil {
+		return err
+	}
+
+	if bcrypt.CompareHashAndPassword(client.Spec.ClientSecretHash, []byte(req.FormValue("client_secret"))) != nil {
+		return types.NewErrHTTP(http.StatusUnauthorized, "Invalid client credentials")
+	}
+
+	grantType := req.FormValue("grant_type")
+	if !slices.Contains(h.oauthConfig.GrantTypesSupported, grantType) {
+		return types.NewErrBadRequest("%v", Error{
+			Code:        ErrInvalidRequest,
+			Description: fmt.Sprintf("grant_type must be one of %s, not %s", strings.Join(h.oauthConfig.GrantTypesSupported, ", "), grantType),
+		})
+	}
+
+	if !slices.Contains(client.Spec.Manifest.GrantTypes, grantType) {
+		return types.NewErrBadRequest("%v", Error{
+			Code:        ErrInvalidRequest,
+			Description: "client is not allowed to use authorization_code grant type",
```

Comment:
```suggestion
			Description: fmt.Sprintf("client is not allowed to use %s grant type", grantType),
```

Reply:
This flow only allows `authorization_code`.